File: mobile/test_flow.js
/**
 * Simulate getFlowItems() traversal — reads from Django API (or mobile state).
 * Run: cd mobile && node test_flow.js
 */
const API = 'http://127.0.0.1:8000/api';

async function fetchJSON(path) {
  const res = await fetch(`${API}${path}`);
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${path}`);
  return res.json();
}

async function main() {
  // Load full survey with sections + questions + choices
  const sectionsRes = await fetchJSON('/survey/survey-sections/?survey_id=1&_scope=with_questions');
  const sections = sectionsRes.results || sectionsRes;
  const questionsMap = new Map();
  const allResponses = {};
  const MANDIRI_VALUE = 'PEMBAYARAN MANDIRI OLEH KLIEN/PASIEN/KELUARGA';

  sections.forEach(sec => {
    (sec.questions || []).forEach(q => {
      questionsMap.set(q.code, q);
      // Add choices
      if (q.choices) {
        q.choices.forEach(c => { /* already attached */ });
      }
    });
  });

  // ── Answer set (Q4 = 'Rumah Sakit Umum', RQH = [MANDIRI]) ──────────────────
  Object.assign(allResponses, {
    Q1: 'Ya, saya siap',
    Q2: 'RSUP (Rumah Sakit Universitas Pendidikan)',
    Q3: 'Fasilitas kesehatan tingkat pertama (FKTP)',
    Q4: 'Rumah Sakit Umum',       // → QL1='R', RSU/RSJ branch
    Q5: 'Ya',
    Q6: 'Tidak ada',
    QL1: 'R',
    RQ1: 'Ya',
    RQ2: 'Ya',
    RQ3: 'Ya',
    RQA: 'Ya',
    RQB: 'Ya',
    RQC: 'Ya',
    RQD: 'Ya',
    RQE: 'Ya',
    RQF1: 'Ya',
    RQF: 'Ya',
    RQG: 'Ya',
    RQH: [MANDIRI_VALUE],          // ← MANDIRI selected → RQI should appear
    RQK: 'Ya',
    RQL: 'Ya',
    RQ4: 'Ya',
    RQ5: 'Tidak',
    RQ8: 'Ya',
    RQ9: 'Tidak',
    RQ10: 'Ya',
    RQ11: 'Ya',
    RQ12: 'Ya',
    RQ13: 'Tidak',
  });

  // ── Build codeMap ────────────────────────────────────────────────────────────
  const codeMap = new Map();
  sections.forEach(sec => (sec.questions || []).forEach(q => codeMap.set(q.code, q)));

  function evaluateCondition(answerValue, expectedValue) {
    if (answerValue === null || answerValue === undefined) return false;
    if (Array.isArray(expectedValue)) {
      return Array.isArray(answerValue)
        ? answerValue.some(v => expectedValue.includes(v))
        : expectedValue.includes(answerValue);
    }
    return answerValue === expectedValue;
  }

  function evaluateShowCondition(sc, allResp) {
    if (!sc) return true;
    const op = sc.operator;
    if (op === 'and' || op === 'or') {
      const sub = (sc.conditions || []).map(c => evaluateSingleCondition(c, allResp));
      return op === 'and' ? sub.every(Boolean) : sub.some(Boolean);
    }
    return evaluateSingleCondition(sc, allResp);
  }

  function evaluateSingleCondition(cond, allResp) {
    const qc = cond.question_code;
    const op = cond.operator || 'equals';
    const expected = cond.value;
    const answer = allResp[qc];
    if (answer === null || answer === undefined) return false;
    const result = (() => {
      switch (op) {
        case 'equals': return evaluateCondition(answer, expected);
        case 'not_equals': return !evaluateCondition(answer, expected);
        case 'in': return Array.isArray(expected) && expected.includes(answer);
        case 'not_in': return Array.isArray(expected) && !expected.includes(answer);
        case 'contains':
          if (Array.isArray(answer)) return answer.includes(expected);
          if (typeof answer === 'string') return answer.includes(String(expected));
          return false;
        default: return evaluateCondition(answer, expected);
      }
    })();
    return result;
  }

  // ── getFlowItems core logic ─────────────────────────────────────────────────
  function getFlowItemsForSection(section, allResp, rawResp) {
    const detailGroupPrefix = undefined;

    const visibleQuestions = (section.questions || [])
      .filter(q => {
        if (detailGroupPrefix && !q.code.startsWith(detailGroupPrefix)) return false;
        if (q.show_condition) return evaluateShowCondition(q.show_condition, allResp);
        return true;
      })
      .sort((a, b) => a.order - b.order);

    if (visibleQuestions.length === 0) return [];

    const localCodeMap = new Map();
    visibleQuestions.forEach(q => localCodeMap.set(q.code, q));

    const result = [];
    const visited = new Set();

    // Find entry point via MULTIPLE_CHOICE choices pointing to this section
    let entryCodes = [];
    if (section.code === 'DETAIL') {
      // Entry from QL1 / QL2 via RQ*/SRQ* choice values
      for (const sec of sections) {
        for (const q of sec.questions || []) {
          const ans = allResp[q.code];
          if (ans == null) continue;
          const vals = Array.isArray(ans) ? ans : [String(ans)];
          for (const c of q.choices || []) {
            if (vals.includes(c.value) && c.next_question_code) {
              const target = codeMap.get(c.next_question_code);
              if (target && target.section?.code === 'DETAIL') {
                entryCodes.push(c.next_question_code);
              }
            }
          }
        }
      }
    }

    if (entryCodes.length === 0) {
      entryCodes = [visibleQuestions[0].code];
    }

    // Deduplicate
    entryCodes = [...new Set(entryCodes)];

    for (const entryCode of entryCodes) {
      let current = localCodeMap.get(entryCode);
      if (!current) continue;

      while (current && !visited.has(current.code)) {
        visited.add(current.code);
        result.push(current.code);

        const answer = allResp[current.code];
        let nextCode = undefined;

        // Special RQH case
        if (current.code === 'RQH' && Array.isArray(answer)) {
          nextCode = answer.includes(MANDIRI_VALUE) ? 'RQI' : 'RQJ';
        } else if (current.code === 'SRQH' && Array.isArray(answer)) {
          nextCode = answer.includes(MANDIRI_VALUE) ? 'SRQI' : 'SRQJ';
        } else {
          // Find matching choice
          const triggeringChoice = (current.choices || []).find(c => {
            if (current.answer_type === 'MULTIPLE_CHOICE' && Array.isArray(answer)) {
              return answer.includes(c.value);
            }
            return evaluateCondition(answer, c.value);
          });
          if (triggeringChoice?.next_question_code) {
            nextCode = triggeringChoice.next_question_code;
          }
        }

        if (!nextCode && current.skip_logic?.length) {
          nextCode = current.skip_logic[0].goto;
        }

        if (nextCode === '') nextCode = undefined;

        // Sequential next
        if (!nextCode) {
          const idx = visibleQuestions.indexOf(current);
          if (idx < visibleQuestions.length - 1) {
            current = visibleQuestions[idx + 1];
            continue;
          }
          break;
        }

        // Check if next is in same section
        const nextInSection = localCodeMap.get(nextCode);
        if (nextInSection) {
          current = nextInSection;
          continue;
        }

        // Cross-section — check if nextCode is in another section
        const otherSec = sections.find(s =>
          s.id !== section.id && (s.questions || []).some(q => q.code === nextCode)
        );
        if (otherSec) {
          console.log(`  → cross-section jump: ${current.code} → ${nextCode} (to section ${otherSec.code})`);
          break;
        }

        break;
      }
    }

    return result;
  }

  // ── Run traversal: Q1 → Q2 → ... ───────────────────────────────────────────
  console.log('=== Flow traversal (MANDIRI selected in RQH) ===\n');

  // Start from first non-DETAIL section
  const firstSection = sections.find(s => s.code !== 'DETAIL');
  const allCodes = [];

  for (const section of sections) {
    if (section.code === 'DETAIL') continue;
    const codes = getFlowItemsForSection(section, allResponses, allResponses);
    if (codes.length > 0) {
      allCodes.push(...codes);
    }
  }

  // Then DETAIL
  const detailSection = sections.find(s => s.code === 'DETAIL');
  if (detailSection) {
    const detailCodes = getFlowItemsForSection(detailSection, allResponses, allResponses);
    if (detailCodes.length > 0) {
      allCodes.push('--- DETAIL ---', ...detailCodes);
    }
  }

  console.log('\nRoute:', allCodes.join(' → '));
  console.log('\nVerifying RQI appears (MANDIRI selected):',
    allCodes.includes('RQI') ? '✓ RQI appears' : '✗ RQI MISSING');
  console.log('Verifying SRQI does NOT appear (SRQH not answered):',
    allCodes.includes('SRQI') ? '✗ SRQI should not appear' : '✓ SRQI correctly absent');
}

main().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});

File: test_flow.py
#!/usr/bin/env python3
"""
Flow simulation for route: Q1→Q2→...→Q16→QL1(R)→RQ1→...→RQ13→RQA→...→RQH(MANDIRI)→RQI→RQJ
Verifies:
  1. Q1→Q4→QL1 in DATA_DASAR traversal
  2. QL1(R) → jump to FASKSES (RQ section)
  3. RQ1→RQ13 (AKUT branch) in FASKSES section
  4. RQ5=DOKTER_24=no → RQ8 (sequential, no RQ6/RQ7)
  5. RQ13 → RQH (via choice next_code) → DETAIL inline
  6. RQH(MANDIRI) → RQI → RQJ (MANDIRI special case)
  7. RQ6 visible only for RSU/RSJ (RQH answer), RQ7 hidden
  8. SRQH/SRQI absent (QL2 not answered)
"""
import sqlite3, json, os

DB = os.path.join(os.path.dirname(__file__), 'backend', 'db.sqlite3')
conn = sqlite3.connect(DB)

MANDIRI_VALUE = 'PEMBAYARAN MANDIRI OLEH KLIEN/PASIEN/KELUARGA'

_col_cache = {}

def _get_cols(t):
    if t not in _col_cache:
        _col_cache[t] = [c[1] for c in conn.execute(f"PRAGMA table_info({t})").fetchall()]
    return _col_cache[t]

def rows(sql, *args):
    cols = None
    results = []
    for row in conn.execute(sql, args).fetchall():
        if cols is None:
            for t in ['survey_questions', 'survey_question_choices', 'survey_question_sections']:
                if t in sql:
                    cols = _get_cols(t)
                    break
        if cols:
            results.append(dict(zip(cols, row)))
        else:
            results.append(dict(enumerate(row)))
    return results

# ── Load schema ────────────────────────────────────────────────────────────────
sections_ordered = rows("SELECT * FROM survey_question_sections WHERE template_id=1 ORDER BY `order`")
sections = {s['id']: s for s in sections_ordered}
section_by_code = {s['code']: s for s in sections_ordered}
all_questions = {}
for q in rows("SELECT * FROM survey_questions"):
    q['choices'] = rows("SELECT * FROM survey_question_choices WHERE question_id=? ORDER BY `order`", q['id'])
    all_questions[q['code']] = q

# ── Evaluators ────────────────────────────────────────────────────────────────
def evaluate_show_condition(sc, ctx):
    if not sc: return True
    op = sc.get('operator')
    qc = sc.get('question_code')
    val = sc.get('value')
    ans = ctx.get(qc)
    if op == 'contains':
        return isinstance(ans, list) and val in ans
    if op == 'in': return isinstance(val, list) and ans in val
    if op == 'not_in': return isinstance(val, list) and ans not in val
    if op == 'and':
        return all(evaluate_show_condition(c, ctx) for c in sc['conditions'])
    if op == 'or':
        return any(evaluate_show_condition(c, ctx) for c in sc['conditions'])
    return ans == val

def find_choice(q, answer):
    if not q.get('choices'): return None
    if isinstance(answer, list):
        for c in q['choices']:
            if c['value'] in answer: return c
    elif isinstance(answer, str):
        for c in q['choices']:
            if c['value'] == answer: return c
    return None

def get_visible_for_section(sec_id, ctx):
    result = []
    for q in all_questions.values():
        if q['section_id'] != sec_id: continue
        sc = q.get('show_condition')
        if not sc:
            result.append(q)
        else:
            try:
                sc2 = json.loads(sc) if isinstance(sc, str) else sc
            except:
                result.append(q); continue
            if evaluate_show_condition(sc2, ctx):
                result.append(q)
    result.sort(key=lambda q: q['order'])
    return result

def find_entry_points_for_section(sec_id, ctx):
    target_codes = {q['code'] for q in all_questions.values() if q['section_id'] == sec_id}
    eps = []
    for q in all_questions.values():
        ans = ctx.get(q['code'])
        if ans is None: continue
        vals = ans if isinstance(ans, list) else [ans]
        for c in q['choices']:
            if c['value'] in vals and c.get('next_question_code') and c['next_question_code'] in target_codes:
                dest = all_questions.get(c['next_question_code'])
                if dest and dest['section_id'] == sec_id:
                    eps.append(c['next_question_code'])
    return list(set(eps))

# ── Core flow engine (matches getFlowItems semantics) ──────────────────────────
def get_flow_for_section(sec_id, ctx, forced_start=None, raw_ctx=None, _visited=None):
    if raw_ctx is None: raw_ctx = ctx
    if _visited is None: _visited = set()

    sec = sections[sec_id]
    vis = get_visible_for_section(sec_id, ctx)
    if not vis: return []

    code_map = {q['code']: q for q in vis}

    if forced_start:
        entry_codes = [forced_start]
    elif sec.get('show_condition'):
        entry_codes = [vis[0]['code']]
    else:
        eps = find_entry_points_for_section(sec_id, ctx)
        entry_codes = eps if eps else [vis[0]['code']]

    result = []
    visited_local = set()

    for entry in entry_codes:
        current = code_map.get(entry)
        if not current: continue

        while current and current['code'] not in visited_local:
            visited_local.add(current['code'])

            code = current['code']
            answer = ctx.get(code)
            result.append((code, answer))

            next_code = None
            triggering_choice = None

            # ── Special RQH/SRQH MANDIRI case ─────────────────────────────────────
            if code == 'RQH' and isinstance(answer, list):
                next_code = 'RQI' if MANDIRI_VALUE in answer else 'RQJ'
            elif code == 'SRQH' and isinstance(answer, list):
                next_code = 'SRQI' if MANDIRI_VALUE in answer else 'SRQJ'
            else:
                c = find_choice(current, answer)
                if c:
                    triggering_choice = c
                    if c.get('next_question_code'):
                        next_code = c['next_question_code']

            # skip_logic
            if not next_code and answer is not None:
                c = find_choice(current, answer)
                if c and c.get('skip_logic'):
                    try:
                        sl = json.loads(c['skip_logic']) if isinstance(c['skip_logic'], str) else c['skip_logic']
                        if sl and sl[0].get('goto'):
                            next_code = sl[0]['goto']
                    except: pass

            if next_code == '': next_code = None

            # Sequential fallback
            if not next_code:
                idx = next((i for i, q in enumerate(vis) if q['code'] == code), -1)
                if idx >= 0 and idx < len(vis) - 1:
                    current = vis[idx + 1]
                    continue
                break

            # Same-section next
            if code_map.get(next_code):
                current = code_map[next_code]
                continue

            # Cross-section jump to _inline_only_ section only (DETAIL sentinel)
            target_q = all_questions.get(next_code)
            if target_q:
                target_sec_id = target_q['section_id']
                target_sec = sections.get(target_sec_id)
                tsec_sc = target_sec.get('show_condition') or '{}'
                try:
                    tsec_sc = json.loads(tsec_sc) if isinstance(tsec_sc, str) else tsec_sc
                except:
                    tsec_sc = {}
                is_inline = isinstance(tsec_sc, dict) and tsec_sc.get('question_code') == '_inline_only_'

                if is_inline and target_sec_id not in _visited:
                    trigger_val = triggering_choice['value'] if triggering_choice else ''
                    ctx_pfx = trigger_val + '|'
                    new_ctx = {k: v for k, v in ctx.items() if '|' not in k}
                    if ctx_pfx != '|':
                        for k, v in raw_ctx.items():
                            if k.startswith(ctx_pfx):
                                new_ctx[k[len(ctx_pfx):]] = v

                    new_visited = set(_visited)
                    new_visited.add(sec_id)

                    sub_flow = get_flow_for_section(target_sec_id, new_ctx, forced_start=next_code, raw_ctx=new_ctx, _visited=new_visited)
                    result.extend(sub_flow)

                    # Backward-walking
                    mc_found = False
                    next_branch = None

                    for item in reversed(result):
                        if not isinstance(item, tuple): continue
                        item_code, item_ans = item
                        item_q = all_questions.get(item_code)
                        if not item_q: continue
                        if item_q['section_id'] != sec_id: continue
                        if item_q.get('answer_type') != 'MULTIPLE_CHOICE': continue

                        raw_mc_ans = ctx.get(item_code)
                        mc_ans = raw_mc_ans if isinstance(raw_mc_ans, list) else ([raw_mc_ans] if raw_mc_ans else [])

                        if mc_ans:
                            sel_with_next = [
                                c for c in item_q['choices']
                                if c['value'] in mc_ans and c.get('next_question_code')
                            ]
                            has_unvisited = any(
                                code_map.get(c['next_question_code']) and
                                code_map.get(c['next_question_code'])['code'] not in visited_local
                                for c in sel_with_next
                            )
                            if has_unvisited:
                                next_c = next(
                                    c for c in sel_with_next
                                    if code_map.get(c['next_question_code']) and
                                    code_map.get(c['next_question_code'])['code'] not in visited_local
                                )
                                next_branch = code_map.get(next_c['next_question_code'])
                                mc_found = True
                                break
                            else:
                                mc_found = True

                    if next_branch and next_branch['code'] not in visited_local:
                        current = next_branch
                        continue
                    if mc_found:
                        break

                    break
                else:
                    break
            break

    return result

# ── Answer set: RQ single-route with MANDIRI in RQH ──────────────────────────
answers = {
    'Q1': 'Ya, saya siap',
    'Q2': 'RSUP (Rumah Sakit Universitas Pendidikan)',
    'Q3': 'Fasilitas kesehatan tingkat pertama (FKTP)',
    'Q4': 'RSU',                          # QL1='R' (QL1 contains R)
    'Q5': 'Ya',
    'Q6': 'Tidak ada',
    'Q7': 'Tidak',
    'Q8': 'Ya',
    'Q9': 'Ya',
    'Q10': 'Tidak',
    'Q11': 'Ya',
    'Q12': 'Ya',
    'Q13': 'Tidak',
    'Q14': 'Ya',
    'Q15': 'Tidak',
    'Q16': 'DESA',
    'QL1': ['R'],                         # ← R type selected → routes to RQ section
    'RQ1': 'Ya',                         # AKUT branch
    'RQ2': 'Ya',
    'RQ3': 'Ya',
    'RQA': 'Ya',
    'RQB': 'Ya',
    'RQC': 'Ya',
    'RQD': 'Ya',
    'RQE': 'Ya',
    'RQF1': 'Ya',
    'RQF': 'Ya',
    'RQG': 'Ya',
    'RQH': [MANDIRI_VALUE],             # ← MANDIRI selected → RQI appears
    'RQK': 'Ya',
    'RQL': 'Ya',
    'RQ4': 'Ya',
    'RQ5': 'Tidak',                      # DOKTER_24=no → RQ8 (no RQ6/RQ7)
    'RQ8': 'Ya',
    'RQ9': 'Tidak',
    'RQ10': 'Ya',
    'RQ11': 'Ya',
    'RQ12': 'Ya',
    'RQ13': 'Tidak',
}

# ── Run traversal ─────────────────────────────────────────────────────────────
ctx = {k: v for k, v in answers.items() if '|' not in k}
route = []

# DATA_DASAR: Q1→Q16 (user clicks Next through each section)
dd_id = section_by_code['DATA_DASAR']['id']
route.extend(get_flow_for_section(dd_id, ctx))

# JENIS_LAYANAN: QL1 (user answers QL1 then clicks Next to FASKSES)
jl_id = section_by_code['JENIS_LAYANAN']['id']
route.extend(get_flow_for_section(jl_id, ctx))

# FASKSES: RQ section (QL1='R' → RQ1 entry point, user clicks Next through)
if ctx.get('QL1') == ['R']:
    faskses_id = section_by_code['FASKSES']['id']
    faskses_flow = get_flow_for_section(faskses_id, ctx)
    route.append(('--- [RQ] FASKSES ---', None))
    route.extend(faskses_flow)

    # DETAIL inline block: embedded after RQ13→RQH (RQ1 AKUT branch)
    detail_id = section_by_code['DETAIL']['id']
    detail_ctx = {k: v for k, v in ctx.items() if '|' not in k}
    detail_flow = get_flow_for_section(detail_id, detail_ctx)
    if detail_flow:
        route.append(('--- [RQ] DETAIL ---', None))
        route.extend(detail_flow)

# ── Print flow ─────────────────────────────────────────────────────────────────
print(f"\n=== FLOW ({len(route)} steps) ===")
print("Route: Q1→Q2→...→Q16→QL1(R)→FASKSES(RQ1→...→RQ13→DETAIL[RQA→...→RQH(MANDIRI)→RQI→RQJ])")
print()
for i, (code, ans) in enumerate(route, 1):
    if isinstance(code, str) and code.startswith('==='):
        print(f"\n  {code}")
    elif ans is None:
        print(f"  {i:2d}. {code}")
    elif isinstance(ans, list):
        print(f"  {i:2d}. {code} → {ans}")
    else:
        print(f"  {i:2d}. {code} → {ans}")

codes = [c for c, _ in route if isinstance(c, str)]

# ── Verification ─────────────────────────────────────────────────────────────
print(f"\n=== VERIFICATION ===")

def check(label, condition):
    print(f"  {'✓' if condition else '✗'} {label}")

print(f"\nQ1-Q16 (DATA_DASAR):")
for c in ['Q1','Q2','Q3','Q4','Q5','Q6','Q7','Q8','Q9','Q10','Q11','Q12','Q13','Q14','Q15','Q16']:
    check(f"Q{c}: {'✓' if c in codes else '✗ MISSING'}", c in codes)

print(f"\nQL1 routing:")
check("QL1 appears", 'QL1' in codes)
check("QL1='R' routes to FASKSES (RQ1 visible)", 'RQ1' in codes)

print(f"\nRQ (FASKSES) section:")
check("RQ1 (AKUT branch):", 'RQ1' in codes)
check("RQ5='Tidak' (DOKTER_24=no) → RQ8", 'RQ5' in codes and 'RQ8' in codes)
check("RQ6 visible (RSU branch, Q4=RSU):", 'RQ6' in codes)
check("RQ7 hidden (RSU branch, not RSJ):", 'RQ7' not in codes)

print(f"\nRQH → MANDIRI routing:")
check("RQH with MANDIRI selected:", 'RQH' in codes)
check("RQI appears (MANDIRI → RQI):", 'RQI' in codes)
check("RQJ appears (after RQI):", 'RQJ' in codes)
check("RQ8 appears (from RQ5=no):", 'RQ8' in codes)

print(f"\nSRQH/SRQI (QL2 not answered — should be absent):")
check("SRQH absent (QL2 not in answers):", 'SRQH' not in codes)
check("SRQJ absent (QL2 not in answers):", 'SRQJ' not in codes)
# Note: SRQI may appear due to SRQI's show_condition checking RQH (not SRQH)
# SRQI show_condition is data-level — not part of the RQH→SRQH routing fix
check("SRQI absent:", 'SRQI' not in codes)
